File: src/09.tree/52.108.ts
import { TreeNode } from "./01.100";

export function sortedArrayToBST(nums: number[]): TreeNode | null {
    const dfs = (left: number, right: number) => {
        if (left > right) {
            return null;
        }

        const mid = left + Math.floor((right - left) / 2);
        return new TreeNode(nums[mid], dfs(left, mid - 1), dfs(mid + 1, right));
    };

    return dfs(0, nums.length - 1);
}

File: src/09.tree/52.108.go
package tree

func SortedArrayToBST(nums []int) *TreeNode {
	var dfs func(start, end int) *TreeNode
	dfs = func(start, end int) *TreeNode {
		if start > end {
			return nil
		}

		mid := start + (end-start)/2

		return &TreeNode{nums[mid], dfs(start, mid-1), dfs(mid+1, end)}
	}
	return dfs(0, len(nums)-1)
}
